### DISCUSSION TITLE
Subdivide a mesh according to the surface normal of its elements

### DISCUSSION THREAD

**Discussion by Xtaltec** [pyvista/pyvista#2545] [Q&A] [answered]
Hi,

In Meshmixer there is a very handy function that lets you subdivide a mesh according to the orientation of the triangular elements as shown in the image below.

![image](https://user-images.githubusercontent.com/42573295/166261163-5cdc9c40-657e-4e3c-980c-bef631d69286.png)

I wonder if there is a similar function in PyVista. I have tried working directly with the surface normals of each cell by myself but I have not been successful so far.

Kind regards

**Comment by darikg** [comment]
Don't think there's a dedicated function, but we can break the problem down into two parts:

1) Calculating an angular distance for each cell from a target direction

```python
from numpy import array, newaxis, arccos, pi
from numpy.linalg import norm

target_direction = array([x, y, z])
target_direction /= norm(target_direction)  # unit vector
cos_theta = (mesh.cell_data['Normals'] * target_direction[newaxis]).sum(axis=1)  # dot product
theta = arccos(cos_theta)
mesh.cell_data['angular_dist'] = theta
```

2) Grouping regions of similar angular distance

There's a [connectivity](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyDataFilters.connectivity.html#pyvista.PolyDataFilters.connectivity) filter in pyvista but it doesn't support using scalar data (yet), so we'll have to drop down to VTK.

```
from vtk import vtkPolyDataEdgeConnectivityFilter 
from pyvista.core.filters import _get_output

angle_threshold = 10 * pi / 180  # 10 degrees

# Use vtkPolyDataConnectivityFilter here instead if you wanted to use point-normals instead of face-normals
# Note that AFAICT there's no way to specify which scalar to use, may need to ensure that angular_dist is the only cell data variable
f = vtkPolyDataEdgeConnectivityFilter()
f.SetScalarConnectivity(True)
f.SetScalarRange(0, angle_threshold)
f.SetExtractionModeToAllRegions()
f.SetColorRegions(True)
f.SetInputData(mesh)
f.Update()
mesh1 = _get_output(f)
region_id = mesh1.cell_data['RegionId']
```

Which assigns each group of faces an integer. Annoyingly, the connectivity filter here assigns every face outside the specified scalar range a unique region id, so one extra post-processing step to collapse them:

```python
region_id[theta > angle_threshold] = -1
```

We could also apply a size filter to the regions:

```python
from numpy import unique

for rid in unique(region_id):
    ridx = region_id == rid
    region = mesh1.extract_cells(ridx)
    if region.area < area_threshold:
        region_id[ridx] = -1
```

Then `mesh1.plot(scalars='RegionId')` does what you want

**Comment by darikg** [comment, marked as answer]
For an approach that doesn't require a target direction, but instead uses edges with sharp angles ("feature edges") to segregate groups of cells, we can use vtkPolyDataEdgeConnectivityFilter barrier edges.

```python
from numpy import arange
from pyvista import PolyData

mesh.point_data['orig_point_idx'] = arange(mesh.n_points)
feature_edges = mesh.extract_feature_edges(
    feature_angle=10,
    boundary_edges=False,
    non_manifold_edges=False,
    manifold_edges=False,
    feature_edges=True,
)

# There's some unfortunate API impedance mismatch here, vtkPolyDataEdgeConnectivityFilter wants a barrier edges 
# specified with the same point indices as the original mesh, but extract_feature_edges only keeps points 
# belonging to the feature edges
edges = feature_edges.lines.reshape(-1, 3)
edges[:, 1:] = feature_edges.point_data['orig_point_idx'][edges[:, 1:]]
feature_edges = PolyData(mesh.points, lines=edges.reshape(-1))

f = vtkPolyDataEdgeConnectivityFilter()
f.SetExtractionModeToAllRegions()
f.SetColorRegions(True)
f.SetInputData(mesh)
f.BarrierEdgesOn()
f.SetSourceData(feature_edges)
f.Update()
mesh1 = _get_output(f)
```

**Comment by Xtaltec** [comment]
I had to update VTK to the latest version, but otherwise, it works like a charm. Your second answer is exactly what I was looking for, thank you very much!
